examples/userapi/mpi.ipynb

```diff
@@ -0,0 +1,1521 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prerequisites\n",
+    "\n",
+    "This notebook contains examples which are expected *to be run with exactly 4 MPI processes*; not because they wouldn't work otherwise, but simply because it's what their description assumes. For this, you need to:\n",
+    "\n",
+    "* Install an MPI distribution on your system, such as OpenMPI, MPICH, or Intel MPI (if not already available).\n",
+    "* Install some optional dependencies, including `mpi4py` and `ipyparallel`; from the root Devito directory, run\n",
+    "```\n",
+    "pip install -r requirements-optional.txt\n",
+    "```\n",
+    "* Create an `ipyparallel` MPI profile, by running our simple setup script. From the root directory, run\n",
+    "```\n",
+    "./scripts/create_ipyparallel_mpi_profile.sh\n",
+    "```\n",
+    "\n",
+    "## Launch and connect to an ipyparallel cluster\n",
+    "\n",
+    "We're finally ready to launch an ipyparallel cluster. Open a new terminal and run the following command\n",
+    "```\n",
+    "ipcluster start --profile=mpi -n 4\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Once the engines have started successfully, we can connect to the cluster"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ipyparallel as ipp\n",
+    "c = ipp.Client(profile='mpi')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, to run commands in parallel over the engines, we will use the %px line magic."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[stdout:0] Hi, I'm rank 0.\n",
+      "[stdout:1] Hi, I'm rank 1.\n",
+      "[stdout:2] Hi, I'm rank 2.\n",
+      "[stdout:3] Hi, I'm rank 3.\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%px\n",
+    "from mpi4py import MPI\n",
+    "print(f\"Hi, I'm rank %d.\" % MPI.COMM_WORLD.rank)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Overview of MPI in Devito\n",
+    "\n",
+    "Distributed-memory parallelism via MPI is designed so that users can \"think sequentially\" for as much as possi-\n",
```

Comment:
Is that a POS(S)IX joke or a forgotten dash?

Reply:
nice spot